pkg/object/bunny.go

```diff
@@ -0,0 +1,186 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package object
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"math"
+	"net/url"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	bunnystorage "github.com/l0wl3vel/bunny-storage-go-sdk"
+)
+
+type bunnyClient struct {
+	DefaultObjectStorage
+	client   *bunnystorage.Client
+	endpoint string
+}
+
+// Description of the object storage.
+func (b *bunnyClient) String() string {
+	return fmt.Sprintf("bunny://%v", b.endpoint)
+}
+
+// Limits of the object storage.
+func (b *bunnyClient) Limits() Limits {
+	return Limits{
+		IsSupportMultipartUpload: false,
+		IsSupportUploadPartCopy:  false,
+	}
+}
+
+// Get the data for the given object specified by key.
+func (b *bunnyClient) Get(key string, off int64, limit int64) (io.ReadCloser, error) {
+	if limit == -1 {
+		limit = math.MaxInt64
+	}
+	body, err := b.client.DownloadPartial(key, off, limit+off-1)
+	if err != nil {
+		return nil, err
+	}
+	return io.NopCloser(bytes.NewReader(body)), nil
+}
+
+// Put data read from a reader to an object specified by key.
+func (b *bunnyClient) Put(key string, in io.Reader) error {
+	content, readErr := io.ReadAll(in)
+	if readErr != nil {
+		return readErr
+	}
+	return b.client.Upload(key, content, true)
+}
+
+// Delete a object.
+// Requires a conditional retry, since deleting a directory or file called foo/bar requires two different calls to the Bunny API, which JuiceFS does not do
+// Deleting a directory requires a trailing slash in the key to delete, which JuiceFS does not add to the path, leading to test case failures.
+// We implement a conditional retry here to try deleting the directory if the delete for a key of the passed name fails.
+// Deleting keys that do not exist are expected to not throw an error
+func (b *bunnyClient) Delete(key string) error {
+	logger.Warnf("Delete: %v", key)
+	if err := b.client.Delete(key, false); err != nil {
+		if err.Error() == "Not Found" {
+			logger.Warnf("Failed to delete %v, %v", key, err)
+			return nil
+		}
+		return err
+	}
+	return nil
+}
+
+func (b *bunnyClient) List(prefix, marker, delimiter string, limit int64, followLink bool) ([]Object, error) {
+	if delimiter != "/" {
+		return nil, notSupported
+	}
+	var output []Object
+	var dir = prefix
+	if !strings.HasSuffix(dir, dirSuffix) { // If no Directory list in parent directory
+		dir = path.Dir(dir)
+		if !strings.HasSuffix(dir, dirSuffix) {
+			dir += dirSuffix
+		}
+	} else if marker == "" { // If Directory && no marker: Return prefix directory as well
+		parentPath := path.Dir(path.Dir(prefix))
+		objects, err := b.client.List(parentPath+dirSuffix)
```

Comment:
We just need to add the directory itself here, will list it later. Please follow the example in file.go

Reply:
I am not sure what changes you would like to see. The code here is based on sftp.go, like you suggested before. Could you elaborate please on what about it is incorrect?

Reply:
Here is how sftp implements List:
```
	var objs []Object
	dir := f.path(prefix)
	if !strings.HasSuffix(dir, "/") {
		dir = filepath.Dir(dir)
		if !strings.HasSuffix(dir, dirSuffix) {
			dir += dirSuffix
		}
	} else if marker == "" {
		obj, err := f.Head(prefix)
		if err != nil {
			if os.IsNotExist(err) {
				return nil, nil
			}
			return nil, err
		}
		objs = append(objs, obj)
	}
	infos, err := c.sftpClient.ReadDir(dir)
	if err != nil {
		if os.IsPermission(err) {
			logger.Warnf("skip %s: %s", dir, err)
			return nil, nil
		}
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}

	entries := f.sortByName(c.sftpClient, dir, infos, followLink)
	for _, o := range entries {
		key := o.Key()
		if !strings.HasPrefix(key, prefix) || (marker != "" && key <= marker) {
			continue
		}
		objs = append(objs, o)
		if len(objs) == int(limit) {
			break
		}
	}
```
When the marker is `''`, it try to return the prefix self (could be a directory). So instead of list the parent of prefix, can we just call Head() instead?

Reply:
I am not sure, but listing a prefix in S3 does not only return the virtual prefix "directory", rather all keys that start with that prefix. This would mean that I implemented the correct behavior. Working on this is also really difficult, since the tests for object storage are very incomplete and miss a lot of corner cases around listing objects.